src/Agent.Plugins/PipelineCache/TarUtils.cs

```diff
@@ -0,0 +1,202 @@
+                }
+
+                var output = new List<string>();
+                Task readLines(string prefix, StreamReader reader) => Task.Run(async () =>
+                  {
+                      string line;
+                      while (null != (line = await reader.ReadLineAsync()))
+                      {
+                          lock (output)
+                          {
+                              output.Add($"{prefix}{line}");
+                          }
+                      }
+                  });
+                Task readStdOut = readLines("stdout: ", process.StandardOutput);
+                Task readStdError = readLines("stderr: ", process.StandardError);
+
+                // Our goal is to always have the process ended or killed by the time we exit the function.
+                try
+                {
+                    using (cancellationToken.Register(() => process.Kill()))
+                    {
+                        // readStdOut and readStdError should only fail if the process dies
+                        // processTcs.Task cannot fail as we only call SetResult on processTcs
+                        await Task.WhenAll(readStdOut, readStdError, processTcs.Task);
+                    }
+
+                    int exitCode = await processTcs.Task;
+
+                    if (exitCode == 0)
+                    {
+                        context.Output($"Process exit code: {exitCode}");
+                        foreach (string line in output)
+                        {
+                            context.Output(line);
+                        }
+                    }
+                    else
+                    {
+                        throw new Exception($"Process returned non-zero exit code: {exitCode}");
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Delete archive file.
+                    if (File.Exists(archiveFile))
+                    {
+                        File.Delete(archiveFile);
+                    }
+                    foreach (string line in output)
+                    {
+                        context.Error(line);
+                    }
+                    ExceptionDispatchInfo.Capture(e).Throw();
+                }
+            }
+            return archiveFile;
+        }
+
+        public async Task<bool> DownloadTar(AgentTaskPluginExecutionContext context, DedupManifestArtifactClient dedupManifestClient, DedupIdentifier dedupId, string targetDirectory, CancellationToken cancellationToken)
+        {
+            var processTcs = new TaskCompletionSource<int>();
+
+            using (var cancelSource = new CancellationTokenSource())
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cancelSource.Token))
+            using (var process = new Process())
+            {
+                process.StartInfo.FileName = isWindows ? "7z" : "tar";
```

Comment:
I know we have a gap with some of our hosted VM images not including "7z" by default ... might be worth adding a quick check (Windows only) to make sure "7z" is in the path and executable. If it's not, the logic can fallback and run "tar" instead. This check should be quick ... simply try to launch a "7z" process (no args) and verify it exists cleanly (no exceptions, 0 exit code)

Reply:
having the behavior vary by whether or not 7z is in the path of the agent's process worries me...

---
In reply to: [313982606](https://github.com/microsoft/azure-pipelines-agent/pull/2369#discussion_r313982606) [](ancestors = 313982606)

Reply:
We always store/create a "tar" file ... this just controls whether "7z" or "tar" is used to un-tar the cache contents (7z has proved to be faster on Windows than "tar", so we want to use "7z" if it's available).